E-Commerce template landing template

gurmukhp opened the pull request:
**Preview**
https://amp-ecommerce.firebaseapp.com/templates/e-commerce/landing.amp

E-Commerce landing template including two reusable components:
- Sidebar
- Hero

**Tested on:**
Chrome (latest & previous)
Safari (latest)
Firefox (latest & previous)
iPhone 7+ iOS 10
Pixel XL Android O Chrome latest
Samsung S6 Chrome latest
Nexus 9
iPad Air (iOS 10)

gurmukhp commented on components/navbar/navbar.css:
```diff
@@ -55,8 +55,7 @@
 .ampstart-navbar-trigger {
-  line-height: var(--navbar-height);
-  font-size: var(--h2);
+  height: 24px;
```
Menu icon is no longer a character but instead an icon font. This was because changes to the font color were not adhered to with the Samsung S6. Changing to an SVG will fill:currentColor ensures the icon is not only consistent across all devices but can also be styled.

camelburrito replied:
i think  this is probably because the char was not unicode or utf8 , if we find the right char this should be solve-able easily

gurmukhp replied:
Note: Try to replace with @equiv; and see if Samsung device renders correctly. http://graphemica.com/%E2%89%A1

gurmukhp commented on css/templates/e-commerce/header.css:
```diff
@@ -12,6 +12,22 @@
 
+.ampstart-navbar-trigger {
+  cursor: pointer;
+  height: 24px;
```
This is hardcoded here since it's the height of the SVG and it allows vertical aligning quite easily.

camelburrito commented on components/navbar/navbar.snip.html:
```diff
@@ -19,10 +19,15 @@
   <header class="ampstart-headerbar fixed flex justify-start items-center top-0 left-0 right-0 pl2 pr4">
-    <div role="button" aria-label="open sidebar" on="tap:{{sidebar-id}}.toggle" tabindex="0" class="ampstart-navbar-trigger {{^sidebaronly}}md-hide lg-hide{{/sidebaronly}} pr2">☰</div>
+    <div role="button" aria-label="open sidebar" on="tap:{{sidebar-id}}.toggle" tabindex="0" class="ampstart-navbar-trigger {{^sidebaronly}}md-hide lg-hide{{/sidebaronly}} pr2">
+      <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
```
- this svg has a weird path when its focused we need to make it square  
- we use svgo to optimize
- can you put it in a separate file and inline it by including it here 
- also now that i think about this, changing/adjusting the image becomes very problematic with this approach. I think we should retain the css/unicode char and find a better UTF8 char. 
- If you still insist on the svg image make fixes 1 ,2,3

gurmukhp replied:
- Path fixed, focus is now a square
- Whitespace removed
- Inlined from img/icons folder
- Keeping SVG in as discussed on ping earlier.

camelburrito commented on components/navbar/sidebar.snip.html:
```diff
@@ -17,44 +17,46 @@
-      {{/nav}}
-    </ul>
-  </nav>
+  <div class="ampstart-sidebar-container">
```
we should not be needing this container. you should be able to do this without the container.

gurmukhp replied:
Good shout! We had it in initially because we were aligning the whole container to the bottom but the design changed. Removed container and related css.

camelburrito commented on components/navbar/sidebar.snip.html:
```diff
@@ -17,44 +17,46 @@
+    </nav>
 
-  {{> ../social-follow/social-follow.snip.html}}
+    {{> ../social-follow/social-follow.snip.html}}
```
drop spaces

gurmukhp replied:
done

camelburrito commented on components/social-follow/social-follow.snip.html:
```diff
@@ -18,7 +18,7 @@
 
 {{#social-follow}}
-<ul class="ampstart-social-follow list-reset flex justify-around items-center flex-wrap m0 mb4">
+<ul class="ampstart-social-follow list-reset flex justify-around items-center flex-wrap {{#classes}}{{classes}}{{/classes}}{{^classes}}m0 mb4{{/classes}}">
```
since you are defaulting this - you should call them {{spacing-classes}}

gurmukhp replied:
done

camelburrito commented on components/hero/hero.css:
```diff
@@ -0,0 +1,65 @@
+
+@media (--breakpoint-lg) {
+  .ampstart-hero-content {
+    width: 30%;
```
this could be done with bass grids (remember you can also use the same grid classes in a flex container.)

gurmukhp replied:
Simplified design by removing 30% & 45% widths and setting a fixed width of 500px to parent container.

camelburrito commented on components/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+{{#hero}}
```
Discussed this with few others , i dont think we will be using a component for this , lets move it to the template for now and call it 

`<template-namepsace>-hero-item `

gurmukhp replied:
Moved to partials folder inside templates and updated templates path to ensure they are not built to /dist/ and therefore not validated.

camelburrito commented on components/hero/hero.css:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  .ampstart-hero-body {
+    max-width: none;
```
you should not need this rule

gurmukhp replied:
Removed and replaced with margin reset.

camelburrito commented on components/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+    <div class="ampstart-hero-img">
+      {{#images}}
+        {{#image}}
+          <amp-img src="{{url}}" width="{{width}}" height="{{height}}" layout="responsive" media="{{media}}" alt="{{alt}}"></amp-img>
```
you should use an SRC set to point to normal and 2x images for retina support, also look at the sizes attribute https://www.ampproject.org/docs/guides/responsive/art_direction , this could also be of interest to you

gurmukhp replied:
Thanks for the tip, it's great that we can define multiple images for each device. Since wide and narrow have different dimensions, I was following the same approach as outlined in [fullpage-hero](https://github.com/ampproject/ampstart/blob/master/components/images/fullpage-hero.snip.html#L22).

Should we also look to update that example or do them together in a separate PR?

camelburrito commented on components/hero/hero.css:
```diff
@@ -0,0 +1,65 @@
+  .ampstart-hero-content {
+    width: 30%;
+  }
+}
```
a better stratergy for your layout will be to use bass flex boxes with grids - http://basscss.com/#basscss-flexbox  - 

use media query to toggle image from absolute (widescreen) (lower zindex)to relative (mobile)

the text will just be on top by DOM flow and you can toggle the columns  on the text between 12 (for mobile) may be 6 or 5(for desktop)

gurmukhp replied:
As discussed on ping, the above approach wouldn't work since ampstart-body has to toggle between absolute and relative.

gurmukhp commented on css/templates/e-commerce/page-vars.css:
```diff
@@ -0,0 +1,38 @@
+ */
+
+:root {
+  --color-primary: #FF8888;
```
Done

gurmukhp commented on css/templates/e-commerce/header.css:
```diff
@@ -0,0 +1,126 @@
+@import 'page-vars';
+
+.ampstart-icon {
+  fill: var(--color-primary);
```
Moved to color-icon, defined a base value as color-accent and overridden in ecommerce template as color-primary

gurmukhp commented on templates/e-commerce/landing.amp.json:
```diff
@@ -0,0 +1,219 @@
+{
+  "home": {
+    "head-tag": {
+      "title": "branding",
```
Renamed to e-commerce until we have a  better name for this template.

gurmukhp commented on css/templates/e-commerce/elements.css:
```diff
@@ -0,0 +1,15 @@
+@import '../../core';
+@import 'page-vars';
```
Can we look to get a solution soon? Just with this one page the CSS is over 200 lines of code.

gurmukhp commented on css/templates/e-commerce/elements.css:
```diff
@@ -0,0 +1,15 @@
+@import 'page-vars';
+
+.ampstart-btn-secondary {
+  background-color: var(--color-white);
```
Ack.

gurmukhp commented:
Thanks for looking at the PR, your comments were clear and actionable. I've addressed all the issues and updated the [preview](https://amp-ecommerce.firebaseapp.com/templates/e-commerce/landing.amp) site.

camelburrito commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+
+.ampstart-navbar-trigger {
+  cursor: pointer;
+  height: 24px;
```
i think this is also there in you navbar css change

gurmukhp replied:
Same as above.

camelburrito commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+.ampstart-navbar-trigger {
+  cursor: pointer;
+  height: 24px;
+  color: var(--color-primary);
```
also this , this is the color that gets inherited, this should be dropped too

gurmukhp replied:
Ah good spot! :) Removed

camelburrito commented on css/templates/e-commerce/page-vars.css:
```diff
@@ -0,0 +1,39 @@
+  --narrow-body-font-size: .9rem;
+
+  /* Wide custom properties */
+  --wide-nav-height: 5rem;
```
--wide-nav-item-height?

gurmukhp replied:
done.

camelburrito commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+
+
+/* HERO */
+.e-commerce-ampstart-hero-content {
```
just `commerce-hero-content` (also change elsewhere) ampstart- should not be used for things created inside templates.

In general hero does not make a lot of sense here - i am not sure what you call this in the prototypes. `commerce-branding-content` or some thing that makes sense would be nice, if not now keep in mind, it should be a simple fined and replace.

gurmukhp replied:
Changed to commerce-hero-content although this doesn't feel right. Can be easily renamed when an appropriate name has been found.

camelburrito commented on css/templates/e-commerce/page-vars.css:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+:root {
```
on your page in general i noticed  if you set your primary/seconday color as white and black and accent color as pink - you should be able to get most of the color theme right (without having to override a lot) Try this and see if it works ,  else we can proceed with what you have

camelburrito replied:
Try to align your colors wrt to the inputs and see how much css can be dropped.

gurmukhp replied:
I've changed primary and accent, it now makes more sense semantically but not much CSS was saved. 

I'm not sure what you mean about aligning wrt to inputs? Please could you clarify.

camelburrito commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+  .e-commerce-ampstart-hero-theme-secondary .e-commerce-ampstart-hero-title,
+  .e-commerce-ampstart-hero-theme-secondary .e-commerce-ampstart-hero-body,
+  .e-commerce-ampstart-hero-theme-secondary .e-commerce-ampstart-hero .ampstart-btn {
+    color: var(--color-secondary);
```
can we just do 
```css 
.e-commerce-ampstart-hero-theme-secondary {
  color: var(--color-secondary);
}
```
or does it have to be more specific?

gurmukhp replied:
Simplified.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+{{#hero}}
+  <div class="e-commerce-ampstart-hero center">
```
`<section>`

gurmukhp replied:
done.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+{{#hero}}
+  <div class="e-commerce-ampstart-hero center">
+    <div class="e-commerce-ampstart-hero-img">
```
you should not need this wrapping div , you can add this class directly to the image

gurmukhp replied:
Removed.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+      {{/images}}
+    </div>
+    <div class="e-commerce-ampstart-hero-content e-commerce-ampstart-hero-theme-{{text-theme}} flex justify-center flex-column">
+      <div>
```
drop this div.

gurmukhp replied:
I added this div to ensure that flexbox only vertically centers the contents of this element as a block. Removing this results in the hero content being displayed incorrectly. I couldn't find any basscss classes to help.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+      <h2 class="e-commerce-ampstart-hero-title h3 caps">{{title}}</h2>
+      <p class="e-commerce-ampstart-hero-body">{{body}}</p>
+      {{#link}}
+        <button class="ampstart-btn ampstart-btn-secondary caps">{{text}}</button>
```
if you want this to be a link use the `<a>` tag if this is a button change the json - if it navigates to a new page or different section on the same page then it is a link

gurmukhp replied:
Changed to <a>

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,20 @@
+{{#hero}}
```
lets call this branding or something that makes sense if possible

gurmukhp replied:
I think we should wait until we get a proper name for this component as we've already changed names 4 times and it'd be good to get a finalised name before updating it across the templates.

camelburrito replied:
That makes sense - can you work on a name for this template as well as this component with @spacedino - this is not going to be merge blocking

gurmukhp commented on css/templates/e-commerce/header.css:
```diff
@@ -0,0 +1 @@
+
```
Done

gurmukhp commented on components/social-follow/social-follow.css:
```diff
@@ -15,5 +15,7 @@
 
 .ampstart-icon {
-  fill: var(--color-font-accent);
+  fill: var(--color-icon);
```
Done.

gurmukhp commented on components/social-follow/social-follow-vars.css:
```diff
@@ -0,0 +1,3 @@
+.ampstart-social-follow {
+  --color-icon: var(--color-font-accent);
```
done.

gurmukhp commented on components/navbar/sidebar.snip.html:
```diff
@@ -17,15 +17,15 @@
 
 <!-- Start Sidebar -->
-<amp-sidebar id="{{sidebar-id}}" class="ampstart-sidebar px3 {{^sidebaronly}}md-hide lg-hide{{/sidebaronly}}" layout="nodisplay">
+<amp-sidebar id="{{sidebar-id}}" class="ampstart-sidebar px3 {{^sidebaronly}}md-hide lg-hide{{/sidebaronly}} {{sidebarclasses}}" layout="nodisplay">
```
Sorry I don't fully understand. In previous PRs you mentioned I should try and use BassCSS classes as much as possible which is the reason for adding the center class to this.

I could add more CSS to the page by adding a center style to ampstart-sidebar but I thought the whole point of using BassCSS was to reduce the amount of custom CSS and therefore less CSS on the page?

In most cases it will always be easier to override ampstart-classes but this will lead to larger CSS.

camelburrito replied:
The template you are building is an opnionated template - in which case a slightly larger css is ok. The whole point of this template is to override classes in the baseline - if there is none then use bass css

gurmukhp replied:
Done.

gurmukhp commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+@import '../../core';
+@import 'page-vars';
```
done.

gurmukhp commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+}
+
+main {
+  max-width: none;
```
done.

gurmukhp commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+
+
+/* TYPOGRAPHY */
+.h3 {
```
I'm not sure this would be any more efficient:

```
.ampstart-title-sm {
  font-size: var(--h3);
  line-height: var(--line-height-4);
  letter-spacing: 0.06rem;
}

.h3 {
  color: var(--color-font-title);
  line-height: normal;
  padding-bottom: var(--space-1);
}
```

We would need to override the line height and letter spacing and then add padding and colour. This results in much larger CSS.

gurmukhp commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+  font-weight: var(--bold-font-weight);
+}
+
+.ampstart-hero .ampstart-btn {
```
Merged into previous style.

gurmukhp commented on css/templates/e-commerce/page.css:
```diff
@@ -0,0 +1,216 @@
+}
+
+.ampstart-navbar-trigger {
+  cursor: pointer;
```
It does, but I'm not including that CSS since it contains many styles that I'd need to override.

camelburrito replied:
That could end up being confusing when we maintain a lot of templates. I think it is better to override at this point. (Eventually we would build an uncss solution that would strip off unwanted css on the page + configurator would let you choose which component to include). 

Ideally you should be always overriding from the baseline css.

gurmukhp replied:
Done.

gurmukhp commented:
Thanks for the feedback @camelburrito, I've taken most of your comments on board and asked for further clarification on others.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,18 @@
+    {{/image}}
+  {{/images}}
+  <div class="commerce-hero-content-content commerce-hero-content-theme-{{text-theme}} flex justify-center flex-column">
+    <div>
```
you should not need this div - please drop

gurmukhp replied:
As discussed, will leave this in for now since it's there for flexbox positioning.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,18 @@
+{{#hero}}
+<section class="commerce-hero-content center">
```
commerce-hero-content-wrapper

gurmukhp replied:
Done.

camelburrito commented on templates/e-commerce/partials/hero/hero.snip.html:
```diff
@@ -0,0 +1,18 @@
+      <amp-img src="{{url}}" width="{{width}}" height="{{height}}" layout="responsive" media="{{media}}" alt="{{alt}}"></amp-img>
+    {{/image}}
+  {{/images}}
+  <div class="commerce-hero-content-content commerce-hero-content-theme-{{text-theme}} flex justify-center flex-column">
```
commerce-hero-content. 

Also commerce-hero-content-theme-{{text-theme}}, do you need this or can you use the nth-child pseudo class?

gurmukhp replied:
Renamed.

The theme is dependant on the colour of the background image which can be in any order, so  it's needed as a class.

gurmukhp commented:
I've addressed the final few nits, hopefully this is good to go in, thanks!

camelburrito commented:
Can you send a PR for the Gulp Updates, ill ask @erwinmombay  to take a look.

gurmukhp commented:
Thanks for merging the code in! 

The gulp file is faster since it only really builds and verifies assets related to templates, it doesn't compile stuff in the www folder. I can look into building a new gulp task that is just for local template development, it will be useful going forward for others I'm sure.
